test/Test/Database/LSMTree/Normal/StateMachine.hs

```diff
@@ -0,0 +1,696 @@
+
+import           Control.Exception (SomeException (..))
+import           Control.Monad ((<=<))
+import           Control.Monad.Class.MonadThrow (MonadCatch (..),
+                     MonadThrow (..))
+import           Control.Monad.IOSim (IOSim)
+import           Control.Monad.Reader (ReaderT (..))
+import           Data.Bifunctor (Bifunctor (..))
+import           Data.Maybe (fromJust, mapMaybe)
+import           Data.Set (Set)
+import qualified Data.Set as Set
+import           Data.Typeable (Proxy (..), Typeable, cast)
+import           Data.Word (Word64)
+import qualified Database.LSMTree.Common as SUT
+import qualified Database.LSMTree.Model.Normal.Admin as Model
+import qualified Database.LSMTree.Normal as SUT
+import           System.FS.API (MountPoint (..), SomeHasFS (..))
+import           System.FS.IO (ioHasFS)
+import           System.IO.Temp (createTempDirectory,
+                     getCanonicalTemporaryDirectory)
+import           Test.Database.LSMTree.Normal.StateMachine.Op
+                     (HasBlobRef (getBlobRef), Op (..))
+import qualified Test.QuickCheck as QC
+import           Test.QuickCheck (Arbitrary, Gen)
+import           Test.QuickCheck.StateModel hiding (Var, vars)
+import           Test.QuickCheck.StateModel.Lockstep
+import qualified Test.QuickCheck.StateModel.Lockstep.Defaults as Lockstep.Defaults
+import qualified Test.QuickCheck.StateModel.Lockstep.Run as Lockstep.Run
+import           Test.Tasty (TestTree, testGroup)
+import           Test.Tasty.HUnit (testCase)
+import           Test.Util.Orphans (Wrap (..))
+
+{-------------------------------------------------------------------------------
+  Test tree
+-------------------------------------------------------------------------------}
+
+tests :: TestTree
+tests = testGroup "Normal.StateMachine" [
+      testCase "labelledExamples" $
+        QC.labelledExamples $ Lockstep.Run.tagActions (Proxy @ModelState)
+      -- TODO: enable once we at least have an implementation for SUT sessions.
+    --  , testProperty "propLockstep @IO" $
+    --     Lockstep.Run.runActionsBracket (Proxy @ModelState)
+    --       (createSystemTempDirectory "Lockstep")
+    --       (removeDirectoryRecursive . fst)
+    --       (\r (_, someHasFS) -> SUT.newSession someHasFS (mkFsPath []) >>= runReaderT r)
+    -- , testProperty "propLockstep @IOSim" $
+    --     Lockstep.Run.runActionsBracket (Proxy @ModelState)
+    --       (SomeHasFS <$> simHasFS' MockFS.empty)
+    --       (const (pure ()))
+    --       (\r someHasFS -> SUT.newSession someHasFS (mkFsPath []) >>= runReaderT r)
+    ]
+
+-- TODO: enable once we at least have an implementation for SUT sessions.
+_createSystemTempDirectory ::  [Char] -> IO (FilePath, SomeHasFS IO)
+_createSystemTempDirectory prefix = do
+    systemTempDir <- getCanonicalTemporaryDirectory
+    tempDir <- createTempDirectory systemTempDir prefix
+    pure (tempDir, SomeHasFS $ ioHasFS (MountPoint tempDir))
```

Comment:
`runActionsBracket` from QLS works in the `with` style, but we can't wrap `runActions` in a `with`-style function, because we have to create/delete the directory every time for every generated action sequence. We need the directory creation/deletion to happen inside the property